test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/MockCollectorIntegrationTests.cs

```diff
@@ -0,0 +1,147 @@
+// <copyright file="MockCollectorIntegrationTests.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+#if !NETFRAMEWORK
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading.Tasks;
+using Grpc.Core;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.AspNetCore.TestHost;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Hosting;
+using OpenTelemetry.Metrics;
+using OpenTelemetry.Proto.Collector.Trace.V1;
+using OpenTelemetry.Tests;
+using OpenTelemetry.Trace;
+using Xunit;
+
+namespace OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests;
+
+public sealed class MockCollectorIntegrationTests
+{
+    [Fact]
+    public async Task TestRecoveryAfterFailedExport()
+    {
+        using var host = await new HostBuilder()
+           .ConfigureWebHost(webBuilder => webBuilder
+               .UseTestServer()
+               .ConfigureServices(services =>
+               {
+                   services.AddSingleton(new MockCollectorState());
+                   services.AddGrpc();
+               })
+               .Configure(app =>
+               {
+                   app.UseRouting();
+
+                   app.UseEndpoints(endpoints =>
+                   {
+                       endpoints.MapGet(
+                           "/MockCollector/SetResponseCodes/{responseCodesCsv}",
+                           (MockCollectorState collectorState, string responseCodesCsv) =>
+                           {
+                               var codes = responseCodesCsv.Split(",").Select(x => int.Parse(x)).ToArray();
+                               collectorState.SetStatusCodes(codes);
+                           });
+
+                       endpoints.MapGrpcService<MockTraceService>();
+                   });
+                }))
+           .StartAsync();
+
+        var httpClient = host.GetTestClient();
```

Comment:
can we use a normal httpclient, not the Test "httpclient"?

Reply:
Yep, we could. That's what we do in the Grpc.Net.Client tests https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/test/OpenTelemetry.Instrumentation.Grpc.Tests/GrpcServer.cs#L71.

Do we have a preference? I kinda like the TestServer/TestClient as I think it makes the test code more concise. That is, I think finding an open port, performing the ConfigureKestrel thing, etc. Not a huge deal, but it's extra noise.

Reply:
Ok, you twisted my arm 😄. Still planning on addressing #3393, but this PR is good to go now.

I hard-coded ports for now making sure to use something that doesn't conflict with our other test with a hard-coded port

https://github.com/open-telemetry/opentelemetry-dotnet/blob/be78372696cfdb4358854fdb32b859b778150661/test/OpenTelemetry.Instrumentation.AspNetCore.Tests/InProcServerTests.cs#L55